Add a "reinstall" button when a block type is not found

ryelle opened the pull request:
## Description
If editing a post with a missing block, this PR adds the ability to reinstall it. It uses the Block Directory work to search for a block with the same name, and if found, gives you the option to reinstall the block. Once installed, the broken block is converted back into the original.

Fixes #24759

## How has this been tested?
Only in-browser testing. I had a post with some missing blocks, one of which can be found on the block directory. You can copy this into the code editor to try it out (will install the [Blissful Buttons plugin](https://wordpress.org/plugins/blissful-buttons/))

```html
<!-- wp:sortabrilliant/blissful-buttons {"className":"is-style-offset"} -->
<div class="wp-block-sortabrilliant-blissful-buttons is-style-offset"><a class="wp-blissful-buttons__link" href="https://wordpress.org">This is a button.</a></div>
<!-- /wp:sortabrilliant/blissful-buttons -->
```

## Screenshots
When first loaded, it searches for the block on wp.org. If no matching blocks are found, it only shows the Keep as HTML button (bottom). If it finds a block, it shows both Reinstall and Keep as HTML (top).

<img width="797" alt="Screen Shot 2020-05-25 at 7 51 14 PM" src="https://user-images.githubusercontent.com/541093/82912967-b5311100-9f3b-11ea-97bd-6e60af5c21f9.png">

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ryelle committed:
Block Directory: Add a "reinstall" button when a block type is not found

ryelle committed:
Update package-lock.json

StevenDufresne commented:
I like this idea and I can see it becoming useful in a number of different "missing blocks" use cases.

gziolo commented on packages/block-directory/src/plugins/index.js:
```diff
@@ -2,14 +2,29 @@
+		if ( name !== 'core/missing' ) {
+			return settings;
+		}
+		settings.edit = missingInstallView;
```
It's fine as a temporary approach. However, it could be very hard to maintain in the long run as you would have to keep the implementation in sync with changes applied to the Missing block.

Have you considered using an approach where you could parameterize action items in some way? It could be also done through filters but on the more granular level or through SlotFill approach.

ryelle committed:
Filter the Missing block edit component

MichaelArestad commented:
This seems like a pretty valuable feature. I suspect the missing block notice may have been overlooked in the 5.5 design updates. Currently, it looks like this:

<img width="577" alt="image" src="https://user-images.githubusercontent.com/1123119/90536876-dfa8f980-e139-11ea-9c5f-ea2e434adbdf.png">

Perhaps we can fix up the padding and button styling in this or a follow up PR to bring it closer to the block placeholders:

<img width="576" alt="image" src="https://user-images.githubusercontent.com/1123119/90537036-141cb580-e13a-11ea-83b9-1ebf9de6e2de.png">


I did modify the copy slightly:

> Your site doesn’t include support for the "%s" block. You can try installing the block, convert it to a Custom HTML block, or remove it entirely.

> Install "%s"

> Keep as HTML

I wasn't able to get the button to appear locally. I tried with Layout Grid and Gutenslider (which I will not be installing again). Hamze has closed the Blissful Buttons plugin so I tested with Layout Grid.

I would like to see what happens when I click the install button. Is there an installing/activating state? Is it so fast it just magically works?

ryelle committed:
Remove debug code

ryelle commented:
I've updated the code here to use a filter on the Missing edit view, so that it can fall back to the core missing component when blocks are not found or the block directory is not allowed. The initial view uses the default missing message, once the block directory returns a block it should load in the "Install [block name]" button.

<img width="644" alt="Screen Shot 2020-08-25 at 6 07 09 PM" src="https://user-images.githubusercontent.com/541093/91233177-aa367a00-e6fe-11ea-83a4-2430f6740599.png">

When clicked, it flips to a loading button while the plugin installs

<img width="663" alt="Screen Shot 2020-08-25 at 6 07 23 PM" src="https://user-images.githubusercontent.com/541093/91233175-a99de380-e6fe-11ea-8b64-9d03bcc03f8e.png">

One thing that still needs work is that the attributes aren't carried over, for example in the above screenshots I've added a "Custom heading label", but when the Table of Contents block installs, it wipes out that content & is like a new block.

StevenDufresne reviewed (commented):
The code looks good. 

Can we get the button to left align with the paragraph content and add the same amount of padding/margin under the button?

ryelle commented on packages/block-directory/src/plugins/missing-install-view/index.js:
```diff
@@ -0,0 +1,105 @@
+			/* translators: %s: block name */
+			__(
+				'Your site doesn’t include support for the "%s" block. You can try reinstalling the block, convert its content to a Custom HTML block, or remove it entirely.'
+			),
```
The current "missing" block uses the same pattern, so I'd rather leave it the same here.

![Screen Shot 2020-08-26 at 10 45 07 AM](https://user-images.githubusercontent.com/541093/91318437-38efd900-e789-11ea-8090-ae1dcd320d27.png)

ryelle commented:
@StevenDufresne Something went weird in your review, it looks like GitHub let you review the old version of this PR. I think the current iteration fixes your other points.

StevenDufresne commented:
> @StevenDufresne Something went weird in your review, it looks like GitHub let you review the old version of this PR. I think the current iteration fixes your other points.

I think i may have had added a comment to a review a while ago but never submitted the review. GitHub must hold on to those, and on my end, it looks like they have been submitted but they actually haven't.

MichaelArestad committed:
Updated warning styles to be more consistent with other patterns like the block placeholder state

* Removed font size declaration - was inconsistent with other patterns particularly because of the use of `em` units for padding
* changed padding on the warning and changed margin on the warning message
* changed the side the padding was on for warning buttons

MichaelArestad commented:
I updated the styles to match the mockup and other patterns (like the block placeholder state). In doing so, I noticed a couple things that were confusing to me. Here's a quick screencap of me disabling a block in order to view the warning:

![2020-08-27 10 47 22](https://user-images.githubusercontent.com/1123119/91471845-d6174400-e853-11ea-8f70-cdb131121dfc.gif)

You'll notice a few things. Once I clicked the button to reinstall Layout Grid, I got a new layout grid essentially. If I then refresh the page, I get my old content back along with the "newer version of this document" warning at the top. Do you know why my original content inside Layout Grid wasn't immediately restored?

The other really minor bug is that the buttons change once the warning figures out there is a block it can install. Is there any way we can prevent that button change flash? I'm open to ideas. This is what I'm talking about:

![2020-08-27 11 01 30](https://user-images.githubusercontent.com/1123119/91472551-dc59f000-e854-11ea-9e2f-3926134c1c90.gif)

ryelle committed:
Run `parse` to get the correct attributes and innerBlocks

ryelle commented:
> The other really minor bug is that the buttons change once the warning figures out there is a block it can install. Is there any way we can prevent that button change flash?

It shows the current missing block view while we're searching the block directory for this block. We don't know if the block exists in the directory (or if it's some custom block), and we don't know the human-friendly name, until the API result comes back. We could add a spinner, or prevent showing any buttons while the search is running, use different copy? I'm not sure what would be best.

> Do you know why my original content inside Layout Grid wasn't immediately restored?

Figured it out! It looks like nested blocks weren't being correctly updated, but they should be now.

MichaelArestad commented:
> We could add a spinner

I think something like that could work. Let's shelve that for a future PR. This one looks good to go from a design standpoint.

StevenDufresne commented on packages/block-editor/src/components/warning/style.scss:
```diff
@@ -3,8 +3,7 @@
 	font-family: $default-font;
-	font-size: $default-font-size;
-	padding: ($grid-unit-10 - $border-width - $border-width) $grid-unit-15;
+	padding: 1em;
```
Having padding set as an `em` unit could lead to some odd outcomes should the block be broken inside of a parent block. It will inherit its font size. May be better to use `rem` units.

MichaelArestad replied:
There is an open PR to switch to em units and the block Placeholder state uses em units. That's the reason for the change. Might as well be consistent.

StevenDufresne replied:
Thanks for the heads up....

 I wonder if that PR is considering third party blocks. This may get interesting. 🍿

StevenDufresne reviewed (approved):
Minus my comment about using `em` units. This works as expected and is good to go.

As @MichaelArestad has mentioned, we will want to address the UI change when we find a block directory block in the future.

mcsf commented:
I'm late to the party, but have you considered using the [existing `setUnregisteredTypeHandlerName` interface](https://github.com/WordPress/gutenberg/blob/2dd0fb4415bbc7dd0cae96f09504cd9bae0c5eef/packages/block-library/src/index.js#L173) to oust `core/missing` rather than introducing the `editor.missingEdit` filter? I'm finding it hard to justify the extra hook (it's fair to say that over time we've been favouring JS hooks less and less when the goal is to extend visual components — typically better solutions are offered either by Slot/Fill or by replacing components, as is the case with `setUnregisteredTypeHandlerName`).

ryelle commented:
Wouldn't `setUnregisteredTypeHandlerName` require me to rewrite all of `core/missing` again? It seems like I'd need to create an entire new missing block to set the handler to. Unless I'm misreading how `setUnregisteredTypeHandlerName` works?

I thought about a Slot/Fill approach, but that would probably look a lot like this filter, since I'd want to fall through to `core/missing`'s view _unless_ we find the block in the directory.

youknowriad commented:
If I'm understanding the issue, the requirements are:

 - Adding UI to the core missing block but doing it as a "plugin" to leave the "block directory" code on its own module (as a plugin).

--
Without introducing the new `editor.missingEdit`, can't we just hook into `blocks.registerBlockType` for the `core/missing` block and override the `edit` property?

mcsf commented:
> Wouldn't `setUnregisteredTypeHandlerName` require me to rewrite all of `core/missing` again? It seems like I'd need to create an entire new missing block to set the handler to. Unless I'm misreading how `setUnregisteredTypeHandlerName` works?
> 
> I thought about a Slot/Fill approach, but that would probably look a lot like this filter, since I'd want to fall through to `core/missing`'s view _unless_ we find the block in the directory.

Yes to all your questions. :) I see the use case for allowing falling back to `core/missing` and wanting to avoid repetition. I still thought I'd ask, since filtering components can become more fragile over time if the decorated component changes. In this particular case, though, it's comforting to see that `filterMissing` is really just a filter — returning either an `OriginalComponent` instance or something entirely different. So I think a hook was a good call.

> Without introducing the new `editor.missingEdit`, can't we just hook into `blocks.registerBlockType` for the `core/missing` block and override the `edit` property?

This does seem worth looking at.

ryelle commented:
> Adding UI to the core missing block but doing it as a "plugin" to leave the "block directory" code on its own module (as a plugin).

Correct. 

> Without introducing the new editor.missingEdit, can't we just hook into blocks.registerBlockType for the core/missing block and override the edit property?

This is actually [exactly what I did in the first pass of the PR,](https://github.com/WordPress/gutenberg/pull/22631/commits/167d48e2d39017c019557b6e4a516429fd64f922) which @gziolo pointed out would be hard to keep in sync with any core `core/missing` changes (https://github.com/WordPress/gutenberg/pull/22631#discussion_r431764268), though that was rewriting more of `core/missing`. I'll try out a PR for a more combined approach using `blocks.registerBlockType`.
